lib/index.js

```diff
@@ -12,125 +12,131 @@ const schema = Joi.object().keys({
+      requestLimit.reset
+    );
 
-      Object.keys(headers).forEach(
-        key => setResponseHeader(request, key, headers[key]));
-    }
-    reply.continue();
-  });
+    Object
+      .keys(headers)
+      .forEach(key => setResponseHeader(request, key, headers[key]));
+  }
+
+  return h.continue;
 }
 
 function getMinimumLimit(limit1, limit2) {
   if (!limit1) { return limit2; }
   if (!limit2) { return limit1; }
 
-  if (limit1 && limit2.remaining > limit1.remaining) {
+  if (limit2.remaining > limit1.remaining) {
     return limit1;
   }
 
   return limit2;
 }
 
-function setupRateLimitEventExt(server, options) {
-  const event = options.event;
-  const extractKey = options.extractKey;
-  const onError = options.onError;
-
-  const extractKeyAndTakeToken = function(limitd, request, reply, type) {
-    extractKey(request, reply, (err, key) =>{
-      if (err) { return reply(err); }
-
-      if (!limitd) {
-        // limitd is not connected, do not fail!
-        return reply.continue();
-      }
-
-      limitd.take(type, key, (err, currentLimitResponse) => {
-        if (err){
-          if (onError) { return onError(err, reply); }
-          // by default we don't fail if limitd is unavailable
-          return reply.continue();
-        }
-
-        const oldMinimumLimitResponse = request.plugins.patova && request.plugins.patova.limit
-        const newMinimumLimitResponse = getMinimumLimit(currentLimitResponse, oldMinimumLimitResponse)
+function buildRateLimitEvent({ extractKey, limitd, onError, type }) {
+  async function getType(request) {
+    if (typeof type !== 'function') {
+      return type;
+    }
 
-        request.plugins.patova = request.plugins.patova || {};
-        request.plugins.patova.limit = newMinimumLimitResponse;
+    try {
+      return await type(request);
+    } catch (err) {
+      throw Boom.wrap(err, 500, 'cannot get bucket type');
+    }
+  }
 
-        if (newMinimumLimitResponse.conformant) {
-          // We continue only if the request is conformat so far
-          return reply.continue();
+  function takeAsync(type, key) {
+    return new Promise(function(resolve, reject) {
+      limitd.take(type, key, (err, limitResponse) => {
+        if (err) {
+          return reject(err);
         }
+        return resolve(limitResponse);
+      });
+    })
+  }
 
-        const error = Boom.tooManyRequests();
-        error.output.headers = new RateLimitHeaders(
-          newMinimumLimitResponse.limit,
-          newMinimumLimitResponse.remaining,
-          newMinimumLimitResponse.reset);
+  const flowControl = { continue: Symbol('CONTINUE') };
 
-        reply(error);
-      });
-    });
-  };
+  return async function onRateLimit(request, h) {
+    if (!limitd) {
+      // limitd is not connected, do not fail!
+      return h.continue;
+    }
 
-  const getType = function(request, reply, callback) {
-    const type = options.type;
+    const bucketType = await getType(request, flowControl);
```

Comment:
Get type does not accepts a second argument

Reply:
Great catch!